src/pku2u/cert_utils/validation.rs

```diff
@@ -2,32 +2,44 @@ use num_bigint_dig::BigUint;
 use picky::key::PublicKey as RsaPublicKey;
 use picky_asn1_x509::signed_data::{CertificateChoices, SignedData};
 use picky_asn1_x509::{Certificate, PublicKey};
+#[cfg(feature = "logging")]
+use tracing::{error, instrument};
 
 use crate::{Error, ErrorKind, Result};
 
 /// validates server's p2p certificate.
 /// If certificate is valid then return its public key.
+#[cfg_attr(feature = "logging", instrument(level = "trace", ret))]
 pub fn validate_server_p2p_certificate(signed_data: &SignedData) -> Result<RsaPublicKey> {
     let certificates = &signed_data.certificates.0 .0;
 
     if let Some(certificate) = certificates.iter().next() {
         let cert: Certificate = match certificate {
             CertificateChoices::Certificate(cert) => picky_asn1_der::from_bytes(&cert.0)?,
-            _ => {
+            _cert => {
+                #[cfg(feature = "logging")]
+                error!("Server sent unsupported certificate format: {:?}", _cert);
+
                 return Err(Error::new(
                     ErrorKind::Pku2uCertFailure,
                     "Received unknown certificate format".into(),
-                ))
+                ));
             }
         };
 
         let public_key = match cert.tbs_certificate.subject_public_key_info.subject_public_key {
             PublicKey::Rsa(rsa) => rsa,
-            _ => {
+            _public_key => {
+                #[cfg(feature = "logging")]
+                error!(
+                    "Server sent unsupported public key type: {:?}. Only RSA keys supported",
+                    _public_key
+                );
```

Comment:
The tracing way:
```suggestion
                error!(
                    public_key: ?_public_key,
                    "Server sent unsupported public key type: {:?}. Only RSA keys supported",
                );
```

Note that you should remove the `_` prefix if you remove the `logging` feature gate.

Reply:
Fixed